WashU-IT-RIS/coldfront-wustl-fork
# Feature/itdev 35118 update data model for billing reporting

Apologies in advance for the size and complexity of this PR.

Manual test cases:
1. Create a prepaid allocation with a prepaid start date in the past, make sure the prepaid time, extends past today
Expected Result: Billing cycle displays as prepaid, prepaid start date, prepaid expiration date, and prepaid time are all visible 
2. Create a prepaid allocation with a prepaid start date that is not the first of the month
Expected Result: An error next to the field that says that the start date must be the first of the month
3. Create a prepaid allocation with a prepaid start date in the future.
Expected Result: Billing cycle displays as monthly, prepaid start date, prepaid expiration date, and prepaid time are all hidden

Additional notes: The calculation of a the prepaid expiration date is handled in a qcluster job so it may not calculate immediately. This qcluster job runs daily, and it also checks the billing_cycles of all the allocations to see if any allocations need to be swapped from prepaid to monthly and vice versa. This job also switches the service_rate to subscription if an allocation is switched to prepaid.
Billing: Prepaid billing runs on the 28th of each month prior to the actual billing month. Ex: Prepaid billing for January will run on the 28th of December. An allocation is included in the prepaid billing report if it's prepaid billing start date is in the billing month of the billing report. Currently the prepaid billing report only handles allocations with prepaid billing date in the future. Prepaid billing dates in the present and past will need to be handled manually. An allocation on the prepaid billing report will show the total cost for all the months up front and then will not appear on a billing report until it becomes monthly or the customer decides to pay for more months.
